Run keystore management tests on docker distros

williamrandolph opened the pull request:
This PR moves some of the Docker-specific setup and teardown from `DockerTests` up to `PackagingTestCase` so that the KeystoreManagement tests will run on Docker. The main thing that needs to happen is setting the testing `Shell` to be a `DockerShell`. Previously, the KeystoreManagementTests were mostly "passing" on Docker because they ran shell commands on the underlying system and then queried a Docker installation. Now, we can be more confident that the tests are doing the right thing in the Docker case.

Fixes #49469

williamrandolph committed:
Add Docker handling to PackagingTestCase

Keystore tests need to be able to run in the Docker case. We can do this
by using a DockerShell instead of a plain Shell when Docker is running.

williamrandolph committed:
Merge branch 'feature-pwd-protected-keystore-2' into docker-keystore-tests

williamrandolph committed:
Refactor cleanup to parent class

elasticmachine commented:
Pinging @elastic/es-core-infra (:Core/Infra/Packaging)

williamrandolph commented:
TODOs remaining:
- [ ] fix TODO for verifying keystore permissions
- [ ] (potentially) remove redundant tests from DockerTests

williamrandolph committed:
Verify permissions on docker keystore

williamrandolph committed:
Remove redundant docker test

williamrandolph committed:
Remove unused import

williamrandolph committed:
Wait for keystore path availability for docker

williamrandolph commented:
@elasticmachine run elasticsearch-ci/packaging-matrix-unix

pugnascotia commented on qa/os/src/test/java/org/elasticsearch/packaging/test/KeystoreManagementTests.java:
```diff
@@ -80,9 +83,30 @@ public void test11InstallPackageDistribution() throws Exception {
+
+        final Installation.Executables bin = installation.executables();
+        Shell.Result r = sh.runIgnoreExitCode(bin.keystoreTool.toString() + " has-passwd");
+        assertThat("has-passwd should fail", r.exitCode, not(is(0)));
```
```suggestion
        assertFalse("has-passwd should fail", r.isSuccess());
```

pugnascotia commented on qa/os/src/test/java/org/elasticsearch/packaging/test/KeystoreManagementTests.java:
```diff
@@ -80,9 +83,30 @@ public void test11InstallPackageDistribution() throws Exception {
+        final Installation.Executables bin = installation.executables();
+        Shell.Result r = sh.runIgnoreExitCode(bin.keystoreTool.toString() + " has-passwd");
+        assertThat("has-passwd should fail", r.exitCode, not(is(0)));
+        assertThat("has-passwd should fail", r.stdout, containsString("ERROR: Keystore is not password-protected"));
```
Since the previous line checked that the command failed, how about changing the message here:

```suggestion
        assertThat("has-passwd didn't fail with the expected error", r.stdout, containsString("ERROR: Keystore is not password-protected"));
```

williamrandolph committed:
Add isDocker shortcut method

williamrandolph committed:
Return full command width from docker ps

williamrandolph committed:
Remove unused import

williamrandolph committed:
Improve ES startup check for docker

Previously we were checking truncated output for the packaged JDK as
an indication that Elasticsearch had started. With new preliminary
password checks, we might get a false positive from ES keystore
commands, so we have to check specifically that the Elasticsearch
class from the Bootstrap package is what's running.

williamrandolph committed:
Respond to PR feedback

rjernst reviewed (approved):
LGTM
